src/Native/Unix/System.Native/pal_networking.c

```diff
@@ -252,82 +252,126 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
 
     entry->CanonicalName = NULL;
     entry->Aliases = NULL;
-    entry->AddressListHandle = info;
-    entry->IPAddressCount = 0;    
+    entry->AddressList = NULL;
+    entry->AddressCount = 0;
 
     // Find the canonical name for this host (if any) and count the number of IP end points.
     for (struct addrinfo* ai = info; ai != NULL; ai = ai->ai_next)
     {
         // If we haven't found a canonical name yet and this addrinfo has one, copy it
         if ((entry->CanonicalName == NULL) && (ai->ai_canonname != NULL))
         {
-            entry->CanonicalName = (uint8_t*)ai->ai_canonname;
+            entry->CanonicalName = (uint8_t*)strdup(ai->ai_canonname);
         }
 
         if (ai->ai_family == AF_INET || ai->ai_family == AF_INET6)
         {
-            entry->IPAddressCount++;
+            entry->AddressCount++;
         }
     }
 
-    return GetAddrInfoErrorFlags_EAI_SUCCESS;
-}
+#if HAVE_GETIFADDRS
+    char name[MAX_HOST_NAME + 1];
+    result = gethostname((char*)name, MAX_HOST_NAME);
+    struct ifaddrs* addrs = NULL;
+    if (result == 0 && strcasecmp((const char*)address, name) == 0)
+    {
+        // Get all interface addresses if the host name corresponds to the local host.
+        result = getifaddrs(&addrs);
 
-static int32_t GetNextIPAddressFromAddrInfo(struct addrinfo** info, IPAddress* endPoint)
-{
-    assert(info != NULL);
-    assert(endPoint != NULL);
+        // If getifaddrs fails, just skip it, the data are no crucial for the result.
+        if (result == 0)
+        {
+            // Count the number of IP end points.
+            for (struct ifaddrs* ifa = addrs; ifa != NULL; ifa = ifa->ifa_next)
+            {
+                if (ifa->ifa_addr == NULL)
+                {
+                    continue;
+                }
+
+                if (ifa->ifa_addr->sa_family == AF_INET || ifa->ifa_addr->sa_family == AF_INET6)
+                {
+                    entry->AddressCount++;
+                }
+            }
+        }
+    }
+#endif
 
-    for (struct addrinfo* ai = *info; ai != NULL; ai = ai->ai_next)
+    if (entry->AddressCount > 0)
     {
-        switch (ai->ai_family)
+        entry->AddressList = (IPAddress*)calloc(entry->AddressCount, sizeof(IPAddress));
+
+        IPAddress* addressList = entry->AddressList;
+
+        for (struct addrinfo* ai = info; ai != NULL; ai = ai->ai_next)
         {
-            case AF_INET:
+            if (ai->ai_family == AF_INET)
             {
                 struct sockaddr_in* inetSockAddr = (struct sockaddr_in*)ai->ai_addr;
 
-                ConvertInAddrToByteArray(endPoint->Address, NUM_BYTES_IN_IPV4_ADDRESS, &inetSockAddr->sin_addr);
-                endPoint->IsIPv6 = 0;
-                break;
+                ConvertInAddrToByteArray(addressList->Address, NUM_BYTES_IN_IPV4_ADDRESS, &inetSockAddr->sin_addr);
+                addressList->IsIPv6 = 0;
+                ++addressList;
+                continue;
```

Comment:
Why use `continue;` in both of these blocks rather than just making the next if into an else if?

Reply:
Because I like the symmetricity of it and I'm not a fan of `if {} else if {}` without final `else {}`. However, if you think that `if {} else if {}` is generally more readable I'll gladly change it.